### PR TITLE
implement harbor project day2 configuration

### PR DESCRIPTION
**Pull request by ChristianLoewel**
I decided to implement this as I need the harbor project day2 configuration for myself.

This adds the ability to create, update and delete Harbor projects via Kubernetes resources, including their metadata, group/user memberships and storage quotas.

I tried to provide unit tests, however I don't have much experience with go tests yet and due to the lack of already existing day2 tests to base mine off of I decided against it.

For the implementation I strongly oriented myself on the existing pull request for robotaccount day2, extending the REST client wrapper and using the HarborServerConfiguration CustomResource.

Currently, the operator plainly deletes a harbor project if a corresponding HarborProject CustomResource was deleted. It may be necessary to implement a way to optionally skip deleting the project when deleting the CustomResource. This would allow disabling the operator management of a project without deleting it and would limit the damage of accidental deletions.

I look forward to your feedback.

Signed-off-by: Christian Löwel <mail@cloewel.de>

### CODE REVIEW

**Review by thcdrt — COMMENTED** [review]
Thank you for your contribution, I have some suggestions and questions

File: controllers/goharbor/project/memberships.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	desiredMemberships, err := createDesiredMemberships(hp.Spec.HarborProjectMemberships)
+	if err != nil {
+		return err
+	}
+
+	// check length of current/desired member arrays, end reconcile if both are empty.
+	currentMembershipsCnt := len(currentMemberships)
+	desiredMembershipsCnt := len(desiredMemberships)
+
+	if currentMembershipsCnt == 0 && desiredMembershipsCnt == 0 {
```
[the remark this answers is not in the document]
Comment (ChristianLoewel) [reply controllers/goharbor/project/memberships.go:73]: That would be for the case of first reconciliation when the previous hash does not yet exist for comparison.

File: apis/goharbor.io/v1beta1/harborproject_types.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec HarborProjectSpec `json:"spec,omitempty"`
+
+	Status HarborProjectStatus `json:"status,omitempty"`
+}
+
+// HarborProjectSpec defines the spec of HarborProject.
+type HarborProjectSpec struct {
```
[the remark this answers is not in the document]
Comment (ChristianLoewel) [reply apis/goharbor.io/v1beta1/harborproject_types.go:34]: Good point, would definitely make sense to manage this via operator (as well as labels). I will take a look at that!

File: controllers/goharbor/project/memberships.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		}
+
+		newMember.RoleID = int64(memberRoleMapping[definedMember.Role])
+		desiredMembers = append(desiredMembers, newMember)
+	}
+
+	return desiredMembers, nil
+}
+
+// marshal all objects into json and hash it.
```
[the remark this answers is not in the document]
Comment (thcdrt) [reply controllers/goharbor/project/memberships.go]: Yes thank you

File: controllers/goharbor/project/memberships.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thcdrt) [reply controllers/goharbor/project/memberships.go:73]: If previousHash is zero value it would be necessary different from currentHash, am I wrong ?

**Review by thcdrt — APPROVED** [review]

**Comment by thcdrt** [comment]
@ChristianLoewel can you resolve conflicts please ?

**Comment by ChristianLoewel** [comment]
Done! Will create a new pull request in a few weeks with tag retention policy, tag immutability, labels.

File: apis/goharbor.io/v1beta1/harborproject_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChristianLoewel) [reply apis/goharbor.io/v1beta1/harborproject_types.go:34]: Sounds good, will create a new PR in a few weeks.

**Comment by thcdrt** [comment]
@MarcelMue Helm charts tests are broken, is it linked to the move of the CRD out of the chart ?

**Comment by MarcelMue** [comment]
> @MarcelMue Helm charts tests are broken, is it linked to the move of the CRD out of the chart ?

Something went wrong when rebasing here. This PR currently adds back the entire CRD into the helm chart. See how it adds 40k lines right now - I believe that is why the test is failing.

**Comment by ChristianLoewel** [comment]
My bad, wasn't aware of the crd change. I removed it now.

**Comment by MarcelMue** [comment]
I am generally very happy with the contribution. I think we should add some testing for the newly added functionalities here (so we test new features in CI). I am happy if this is done in a follow up / collaboration.

Sorry that it took so long to get this PR reviewed / unblocked :bowing_man:

**Comment by MarcelMue** [comment]
@ChristianLoewel can you provide some manifests (e.g. sample yaml config of CRDs) which you used for testing the added features in #1044 ? It will be helpful later on to more easily add tests.

**Comment by ChristianLoewel** [comment]
Hi, sorry for the delay. I have created some test resources, let me know if you miss any cases.

https://gist.github.com/ChristianLoewel/24fc73b8ac740be5d89283a64df15c53

Tests would be great, I like the idea of the collaboration. Will definitely need your help with that.

**Comment by MarcelMue** [comment]
Thanks for the contribution @ChristianLoewel ! If you want to collaborate more, please feel free to join the `#harbor-dev` channel in the cncf slack: https://cloud-native.slack.com/messages/harbor-operator-dev/

**Comment by chlins** [comment]
> Hi, sorry for the delay. I have created some test resources, let me know if you miss any cases.
> 
> https://gist.github.com/ChristianLoewel/24fc73b8ac740be5d89283a64df15c53
> 
> Tests would be great, I like the idea of the collaboration. Will definitely need your help with that.

@ChristianLoewel Hi, could you please paste the example resources to the [documents](https://github.com/goharbor/harbor-operator/blob/main/docs/day2/day2-harborprojects.md)?

**Comment by ChristianLoewel** [comment]
Hi @chlins, I have created a pull request  #1066 to add the examples to the doc.

Additionally, a second PR #1067 fixes a reconciliation issue with HarborProjects I have noticed while using the feature.

**Pull request merged**